drivers/lora/sx1276.c

```diff
@@ -0,0 +1,492 @@
+{
+	irq_unlock(*mask);
+}
+
+uint32_t RtcGetTimerElapsedTime(void)
+{
+	return counter_read(dev_data.counter);
+}
+
+u32_t RtcGetMinimumTimeout(void)
+{
+	/* TODO: Get this value from counter driver */
+	return 3;
+}
+
+void RtcSetAlarm(uint32_t timeout)
+{
+	struct counter_alarm_cfg alarm_cfg;
+
+	alarm_cfg.flags = 0;
+	alarm_cfg.ticks = counter_us_to_ticks(dev_data.counter, timeout);
+
+	counter_set_channel_alarm(dev_data.counter, 0, &alarm_cfg);
+}
+
+uint32_t RtcSetTimerContext(void)
+{
+	return 0;
+}
+
+uint32_t RtcMs2Tick(uint32_t milliseconds)
+{
+	return counter_us_to_ticks(dev_data.counter, (milliseconds / 1000));
+}
+
+void DelayMsMcu(uint32_t ms)
+{
+	k_sleep(ms);
+}
+
+static void sx1276_irq_callback(struct device *dev,
+				struct gpio_callback *cb, u32_t pins)
+{
+	unsigned int i, pin;
+
+	pin = find_lsb_set(pins) - 1;
+
+	for (i = 0; i < SX1276_MAX_DIO; i++) {
+		if (pin == sx1276_dio_pins[i]) {
+			(*DioIrq[i])(NULL);
+		}
+	}
+}
+
+void SX1276IoIrqInit(DioIrqHandler **irqHandlers)
+{
+	unsigned int i;
+	static struct gpio_callback callbacks[SX1276_MAX_DIO];
+
+	/* Setup DIO gpios */
+	for (i = 0; i < SX1276_MAX_DIO; i++) {
+		dev_data.dio_dev[i] = device_get_binding(sx1276_dio_ports[i]);
+		if (dev_data.dio_dev[i] == NULL) {
+			LOG_ERR("Cannot get pointer to %s device",
+				sx1276_dio_ports[i]);
+			return;
+		}
+
+		gpio_pin_configure(dev_data.dio_dev[i], sx1276_dio_pins[i],
+				   GPIO_DIR_IN | GPIO_INT | GPIO_INT_EDGE |
+				   GPIO_INT_DEBOUNCE | GPIO_INT_ACTIVE_HIGH);
+
+		gpio_init_callback(&callbacks[i],
+				   sx1276_irq_callback,
+				   BIT(sx1276_dio_pins[i]));
+
+		if (gpio_add_callback(dev_data.dio_dev[i], &callbacks[i]) < 0) {
+			LOG_ERR("Could not set gpio callback.");
+			return;
+		}
+		gpio_pin_enable_callback(dev_data.dio_dev[i],
+					 sx1276_dio_pins[i]);
+	}
+
+}
+
+static int sx1276_transceive(u8_t reg, bool write, void *data, size_t length)
+{
+	const struct spi_buf buf[2] = {
+		{
+			.buf = &reg,
+			.len = 1
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 2,
+	};
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
+		};
+
+		return spi_transceive(dev_data.spi, &dev_data.spi_cfg,
+				&tx, &rx);
+	}
+
+	return spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+}
+
+int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(reg_addr, false, data, len);
+}
+
+int sx1276_write(u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+}
+
+void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_write(addr, buffer[i]);
+		if (ret < 0) {
+			LOG_ERR("Unable to read address: %x", addr);
+			return;
+		}
+	}
+}
+
+void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+	u8_t regval;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_read(addr, &regval, 1);
```

Comment:
Is there a reason why you are not using a burst read?

Reply:
This is not I2C! Is there something like this available for SPI too?

Reply:
Currently sx1276_write() and sx1276_read() operate by 1 byte.
I've only modified sx1276_read() like this:
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index ae2e854..3b1cdfb 100644
--- a/drivers/lora/sx1276.c
+++ b/drivers/lora/sx1276.c
@@ -241,17 +241,12 @@ void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
 
 void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
 {
-       int ret, i;
-       u8_t regval;
-
-       for (i = 0; i < size; i++) {
-               ret = sx1276_read(addr, &regval, 1);
-               if (ret < 0) {
-                       LOG_ERR("Unable to read address: %x", addr);
-                       return;
-               }
+       int ret;
 
-               buffer[i] = regval;
+       ret = sx1276_read(addr, buffer, size);
+       if (ret < 0) {
+               LOG_ERR("Unable to read address: %x", addr);
+               return;
        }
 }
```

The figure below shows 10 bytes read of FIFO register.

**before**
![before](https://user-images.githubusercontent.com/10510127/67536617-f695df80-f712-11e9-894e-ff070d6d251c.png)

**after**
![after](https://user-images.githubusercontent.com/10510127/67536622-0281a180-f713-11e9-8ce5-745a63b77b14.png)

Reply:
@Mani-Sadhasivam you are still not using the burst mode. Is there a reason for that?

Reply:
@alexanderwachter Ah, I overlooked it. Will do.

Reply:
@alexanderwachter After doing burst read, the received data looks incorrect:

```
[00:00:01.129,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:03.135,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:05.141,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:07.147,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:09.153,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:11.159,000] <inf> lora_receive: Received data: hdmlnwosld
```
@KwonTae-young Did you also see the issue?

Reply:
@Mani-Sadhasivam Try adding `STM32_OSPEEDR_VERY_HIGH_SPEED` to the SPI pin configuration.

I tested it at https://github.com/KwonTae-young/zephyr/tree/lora_burst and it works fine.

``` diff
diff --git a/boards/arm/96b_wistrio/pinmux.c b/boards/arm/96b_wistrio/pinmux.c
index 12a11a30af14..19174cd0d4a4 100644
--- a/boards/arm/96b_wistrio/pinmux.c
+++ b/boards/arm/96b_wistrio/pinmux.c
@@ -27,9 +27,9 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB9, STM32L1X_PINMUX_FUNC_PB9_I2C1_SDA},
 #endif /* CONFIG_I2C_1 */
 #ifdef CONFIG_SPI_1
-	{STM32_PIN_PA5, STM32L1X_PINMUX_FUNC_PA5_SPI1_SCK},
-	{STM32_PIN_PA6, STM32L1X_PINMUX_FUNC_PA6_SPI1_MISO},
-	{STM32_PIN_PA7, STM32L1X_PINMUX_FUNC_PA7_SPI1_MOSI},
+	{STM32_PIN_PA5, STM32L1X_PINMUX_FUNC_PA5_SPI1_SCK | STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PA6, STM32L1X_PINMUX_FUNC_PA6_SPI1_MISO | STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PA7, STM32L1X_PINMUX_FUNC_PA7_SPI1_MOSI | STM32_OSPEEDR_VERY_HIGH_SPEED},
 #endif /* CONFIG_SPI_1 */
 	/* RF_CTX_PA */
 	{STM32_PIN_PA4, STM32_PUSHPULL_PULLUP},
```

Reply:
@KwonTae-young Great! It works :) Btw, we only need the flag for SCLK.

Reply:
@Mani-Sadhasivam
I think we also need to add a burst write.
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index d779025..5936e0f 100644
--- a/zephyrproject_lora3/zephyr/drivers/lora/sx1276.c
+++ b/zephyrproject_lora4/zephyr/drivers/lora/sx1276.c
@@ -221,21 +221,18 @@ int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
 	return sx1276_transceive(reg_addr, false, data, len);
 }

-int sx1276_write(u8_t reg_addr, u8_t byte)
+int sx1276_write(u8_t reg_addr, u8_t *data, u8_t size)
 {
-	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+	return sx1276_transceive(reg_addr | BIT(7), true, data, size);
 }

 void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
 {
-	int ret, i;
+	int ret;

-	for (i = 0; i < size; i++) {
-		ret = sx1276_write(addr, buffer[i]);
-		if (ret < 0) {
-			LOG_ERR("Unable to write address: 0x%x", addr);
-			return;
-		}
+	ret = sx1276_write(addr, buffer, size);
+	if (ret < 0) {
+		LOG_ERR("Unable to write address: 0x%x", addr);
 	}
 }

@@ -308,13 +305,13 @@ void SX1276SetRfTxPower(int8_t power)
 	pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
 			     ((power + 1) & 0x0F);
 #endif
-	ret = sx1276_write(SX1276_REG_PA_CONFIG, pa_config);
+	ret = sx1276_write(SX1276_REG_PA_CONFIG, &pa_config, 1);
 	if (ret < 0) {
 		LOG_ERR("Unable to write PA config");
 		return;
 	}

-	ret = sx1276_write(SX1276_REG_PA_DAC, pa_dac);
+	ret = sx1276_write(SX1276_REG_PA_DAC, &pa_dac, 1);
 	if (ret < 0) {
 		LOG_ERR("Unable to write PA dac");
 		return;
```